kms_log/main.tf

```diff
@@ -0,0 +1,398 @@
+                "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
+            ]
+        }
+    }
+
+    statement {
+        sid = "Allow CloudWatch Events Access"
+        effect = "Allow"
+        actions = [
+            "kms:GenerateDataKey",
+            "kms:Decrypt"
+        ]
+        resources = [
+            "*"
+        ]
+        principals {
+            type = "Service"
+            identifiers = [
+                "events.amazonaws.com",
+                "sns.amazonaws.com"
+            ]
+        }
+    }
+}
+
+resource "aws_kms_alias" "kms_logging" {
+    name = "${local.kms_alias}"
+    target_key_id = "${aws_kms_key.kms_logging.key_id}"
+}
+
+# create queue to receive cloudwatch events
+resource "aws_sqs_queue" "dead_letter" {
+    name = "${var.env_name}-kms-dead-letter"
+    kms_master_key_id = "${aws_kms_key.kms_logging.arn}"
+    kms_data_key_reuse_period_seconds = 600
+    message_retention_seconds = 604800 # 7 days
+    tags = {
+        environment = "${var.env_name}"
+    }
+}
+
+resource "aws_sqs_queue" "kms_ct_events" {
+    name = "${var.env_name}-kms-ct-events"
+    delay_seconds = 60
+    max_message_size = 2048
+    visibility_timeout_seconds = 60
+    message_retention_seconds = 345600 # 4 days
+    kms_master_key_id = "${aws_kms_key.kms_logging.arn}"
+    kms_data_key_reuse_period_seconds = 600
+    redrive_policy = <<POLICY
+{
+    "deadLetterTargetArn": "${aws_sqs_queue.dead_letter.arn}",
+    "maxReceiveCount": 10
+}
+POLICY
+tags = {
+        environment = "${var.env_name}"
+    }
+}
+
+resource "aws_sqs_queue_policy" "default"{
+    queue_url = "${aws_sqs_queue.kms_ct_events.id}"
+    policy = "${data.aws_iam_policy_document.sqs_kms_ct_events_policy.json}"
+}
+
+data "aws_iam_policy_document" "sqs_kms_ct_events_policy" {
+    statement {
+        sid = "Allow CloudWatch Events"
+        effect = "Allow"
+        actions = ["sqs:SendMessage"]
+        principals {
+            type = "Service"
+            identifiers = ["events.amazonaws.com"]
+        }
+        resources = ["${aws_sqs_queue.kms_ct_events.arn}"]
+        condition {
+            test = "StringLike"
+            variable = "aws:SourceArn"
+            values = [
+                "${aws_cloudwatch_event_rule.decrypt.arn}"
+            ]
+
+        }
+    }
+}
+
+# cloudwatch event rule to capture decryption events
+resource "aws_cloudwatch_event_rule" "decrypt" {
+    count = "${var.kmslogging_service_enabled}"
+    name = "${var.env_name}-decryption-events"
+    description = "Capture decryption events"
+
+    event_pattern = <<PATTERN
+{
+    "source": [
+        "aws.kms"
+    ],
+    "detail-type": [
+        "AWS API Call via CloudTrail"
+    ],
+    "detail": {
+        "eventSource": [
+            "kms.amazonaws.com"
+        ],
+        "requestParameters": {
+            "encryptionContext": {
+                "context": [
+                    "password-digest",
+                    "pii-encryption"
```

Comment:
From discussion:

These context filters make sure that we're only capturing encrypt/decrypt events that have encryption context with the expected "password-digest" or "pii-encryption" fields.

TODO:
- [ ] Determine whether we should add `session-encryption` once https://github.com/18F/identity-idp/pull/2787 is deployed
- [ ] Determine whether we should monitor for unexpected Encrypt/Decrypt events that lack an encryption context. This may be impossibly noisy if we still have huge numbers of user records encrypted without any context.